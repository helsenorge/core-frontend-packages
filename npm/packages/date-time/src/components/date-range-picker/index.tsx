import * as React from 'react';

import moment from 'moment';
import 'react-dates/initialize';
import { FocusedInputShape, SingleDatePickerShape, DateRangePickerShape } from 'react-dates';

import { LanguageLocales } from '@helsenorge/core-utils/constants/languages';
import { isMobileUA } from '@helsenorge/core-utils/user-agents-utils';

import ValidationError from '@helsenorge/form/components/molecules/form/validation-error';
import {
  DEFAULT_DATE_PLACEHOLDER_NB,
  DEFAULT_DATE_PLACEHOLDER_EN,
  DEFAULT_STARTDATE_PLACEHOLDER,
  DEFAULT_ENDDATE_PLACEHOLDER,
  DEFAULT_DATE_PLACEHOLDER_SE,
} from './../../../constants/datetime';
import { renderDesktopDatePicker } from './date-range-picker-desktop';
import { DateRangePickerLabel } from './date-range-picker-label';
import { renderMobileDatePicker } from './date-range-picker-mobile';
import { DateRangePickerDefaultProps, DateRangePickerProps, DateRangePickerState } from './date-range-picker-types';
import {
  START_DATE_ID,
  END_DATE_ID,
  Phrases,
  ErrorPhrases,
  getCSSClasses,
  getDateNOString,
  notifyValidated,
  getDefaultVisibleMonth,
} from './date-range-picker-utils';
import { validateSingleDate, validateRangeDate, isPrevMonthValid, isNextMonthValid } from './date-range-picker-validation';

import 'react-dates/lib/css/_datepicker.css';
import toolkitstyles from './styles.module.scss';

export class DateRangePicker extends React.Component<DateRangePickerProps, DateRangePickerState> {
  public static defaultProps: DateRangePickerDefaultProps = {
    id: 'daterangepicker',
    resources: Phrases,
    errorResources: ErrorPhrases,
    startDateId: START_DATE_ID,
    startDatePlaceholderText: DEFAULT_STARTDATE_PLACEHOLDER,
    endDateId: END_DATE_ID,
    endDatePlaceholderText: DEFAULT_ENDDATE_PLACEHOLDER,
    showDefaultInputIcon: true,
    isValidationHidden: false,
    isRequired: false,
    isDisabled: false,
    numberOfMonths: 1,
  };

  airbnbSingleDatepickerRef: React.RefObject<React.ClassicComponent<SingleDatePickerShape, {}>>;
  airbnbDateRangepickerRef: React.RefObject<React.ClassicComponent<DateRangePickerShape, {}>>;

  constructor(props: DateRangePickerProps) {
    super(props);

    const momentLocale = moment();
    if (props.locale === LanguageLocales.ENGLISH) {
      require('moment/locale/en-gb');
      momentLocale.locale('en-gb');
    } else if (props.locale === LanguageLocales.SAMI_NORTHERN) {
      require('moment/locale/se');
      momentLocale.locale('se');
    } else {
      require('moment/locale/nb');
      momentLocale.locale('nb');
    }

    this.airbnbSingleDatepickerRef = React.createRef();
    this.airbnbDateRangepickerRef = React.createRef();

    let placeholder = '';

    switch (props.locale) {
      case LanguageLocales.ENGLISH:
        placeholder = DEFAULT_DATE_PLACEHOLDER_EN;
        break;
      case LanguageLocales.SAMI_NORTHERN:
        placeholder = DEFAULT_DATE_PLACEHOLDER_SE;
        break;
      default:
        placeholder = DEFAULT_DATE_PLACEHOLDER_NB;
    }

    this.state = {
      isMobile: isMobileUA(),
      momentLocale: momentLocale.localeData(),
      placeholder: props.placeholder ? props.placeholder : placeholder,
      singleDate: props.singleDateValue ?? null,
      isSingleDateFocused: null,
      isSingleDateValid: true,
      startDate: props.startDateValue ?? null,
      endDate: props.endDateValue ?? null,
      isRangeDateValid: true,
      validated: false,
      focusedInput: null,
      visibleMonth: getDefaultVisibleMonth(props),
    };
  }

  /** Håndterer deaktivering av måned knappene hvis de minimum/maximumDate er satt */
  componentDidMount() {
    this.disableMonthSelectorHandler(this.state.visibleMonth);
  }

  /** Hvis valuen som kommer fra prop er endret underveis så må den gå gjennom validator */
  componentDidUpdate(prevProps: DateRangePickerProps, prevState: DateRangePickerState): void {
    if (prevProps.type === 'single') {
      const prevPropSingleDateValue = getDateNOString(prevProps.singleDateValue);
      const prevStateSingleDate = getDateNOString(prevState.singleDate);
      const currentSingleDateValue = getDateNOString(this.props.singleDateValue);
      if (prevPropSingleDateValue !== currentSingleDateValue && prevStateSingleDate !== currentSingleDateValue) {
        this.setState(
          {
            singleDate: this.props.singleDateValue ?? null,
            visibleMonth: getDefaultVisibleMonth(this.props),
          },
          () => {
            if (!this.state.isSingleDateFocused) {
              this.singleDateValidator(this.state.singleDate);
            }
          }
        );
      }
    } else {
      const prevPropStartDateValue = getDateNOString(prevProps.startDateValue);
      const prevStateStartDate = getDateNOString(prevState.startDate);
      const currentStartDateValue = getDateNOString(this.props.startDateValue);
      const prevPropEndDateValue = getDateNOString(prevProps.endDateValue);
      const prevStateEndDate = getDateNOString(prevState.endDate);
      const currentEndDateValue = getDateNOString(this.props.endDateValue);
      if (prevPropStartDateValue !== currentStartDateValue && prevStateStartDate !== currentStartDateValue) {
        this.setState(
          {
            startDate: this.props.startDateValue ?? null,
          },
          () => {
            if (this.state.focusedInput === null) {
              this.rangeDateValidator(this.state.startDate, this.state.endDate);
            }
          }
        );
      }
      if (prevPropEndDateValue !== currentEndDateValue && prevStateEndDate !== currentEndDateValue) {
        this.setState({ endDate: this.props.endDateValue ?? null }, () => {
          if (this.state.focusedInput === null) {
            this.rangeDateValidator(this.state.startDate, this.state.endDate);
          }
        });
      }
    }

    /* Sjekker om nåværende måned vi ser har endret seg, eller om det skal være lov å gå lenger forover/bakover i tid */
    if (
      prevState.visibleMonth !== this.state.visibleMonth ||
      (prevProps.maximumDate && this.props.maximumDate && !prevProps.maximumDate.isSame(this.props.maximumDate, 'month')) ||
      (prevProps.minimumDate && this.props.minimumDate && !prevProps.minimumDate.isSame(this.props.minimumDate, 'month'))
    ) {
      this.disableMonthSelectorHandler(this.state.visibleMonth);
    }
  }

  /* Calling the validation on the incoming date */
  onSingleDateChange = (singleDate: moment.Moment): void => {
    this.setState({ singleDate });
    this.props.onDateChange && this.props.onDateChange(singleDate, this.props.id);
  };

  /* Calling the validation on the incoming date with the stored end date */
  onStartDateChange = (startDate: moment.Moment): void => {
    this.setState({ startDate });
    this.props.onDateChange && this.props.onDateChange({ start: startDate, end: this.state.endDate }, this.props.id);
  };

  /* Calling the validation on the incoming date with the stored start date */
  onEndDateChange = (endDate: moment.Moment): void => {
    this.setState({ endDate });
    this.props.onDateChange && this.props.onDateChange({ start: this.state.startDate, end: endDate }, this.props.id);
  };

  onRangeDatesChange = ({ startDate, endDate }: { startDate: moment.Moment; endDate: moment.Moment }): void => {
    this.setState({ startDate, endDate });
    this.props.onDateChange && this.props.onDateChange({ start: startDate, end: endDate }, this.props.id);
  };

  singleDateValidator = (date: moment.Moment | null): Promise<void> => {
    const { id, isRequired, errorResources, minimumDate, maximumDate, dateValidator, onValidated, onError } = this.props;

    const v: {
      isSingleDateValid: boolean;
      errorString?: string;
    } = validateSingleDate(this.airbnbSingleDatepickerRef, date, id, isRequired, errorResources, minimumDate, maximumDate, dateValidator);

    return new Promise<void>((resolve: () => void) => {
      this.setState(
        {
          validated: true,
          isSingleDateValid: v.isSingleDateValid,
          errorString: v.errorString,
        },
        () => {
          notifyValidated(id, date, v.isSingleDateValid, v.errorString, onValidated, onError);
          resolve();
        }
      );
    });
  };

  rangeDateValidator = (startDate: moment.Moment | null, endDate: moment.Moment | null): Promise<void> => {
    const { id, isRequired, errorResources, minimumDate, maximumDate, minimumPeriod, dateValidator, onValidated, onError } = this.props;

    const v: {
      isRangeDateValid: boolean;
      errorString?: string;
    } = validateRangeDate(
      this.airbnbDateRangepickerRef,
      startDate,
      endDate,
      id,
      isRequired,
      errorResources,
      minimumDate,
      maximumDate,
      minimumPeriod,
      dateValidator
    );

    return new Promise<void>((resolve: () => void) => {
      this.setState(
        {
          validated: true,
          isRangeDateValid: v.isRangeDateValid,
          errorString: v.errorString,
        },
        () => {
          notifyValidated(id, { start: startDate, end: endDate }, v.isRangeDateValid, v.errorString, onValidated, onError);
          resolve();
        }
      );
    });
  };

  onSingleDateFocusChange = ({ focused }: { focused: boolean | null }): void => {
    // This replaces onBlur method and validates the field for everytime the focus is brought outside
    // setTimeout is needed to delay the check until the focus is moved
    if (!focused) {
      setTimeout(() => {
        this.validateField();
      }, 0);
    }
    if (focused !== this.state.isSingleDateFocused) {
      this.setState({ isSingleDateFocused: focused });
    }
  };

  onRangeDatesFocusChange = (focusedInput: FocusedInputShape): void => {
    // This replaces onBlur method and validates the field for everytime the focus is brought outside
    // setTimeout is needed to delay the check until the focus is moved
    setTimeout(() => {
      if (this.state.focusedInput === null) {
        this.validateField();
        this.setState({ isFocusReset: true });
      }
    }, 0);

    this.setState({ focusedInput });

    // Rekalkulerer om navNext og navPrev skal være disabled
    if (this.state.isFocusReset) {
      this.setState({ isFocusReset: false });
    }
  };

  onCalenderIconClick = (): void => {
    if (this.state.isSingleDateFocused) {
      this.setState({ isSingleDateFocused: false });
    } else if (this.state.focusedInput) {
      this.setState({ focusedInput: null });
    }

    this.setState({ isFocusReset: true });
  };

  /* Runs dates validator */
  validateField(): Promise<void> {
    return this.props.type === 'single'
      ? this.singleDateValidator(this.state.singleDate)
      : this.rangeDateValidator(this.state.startDate, this.state.endDate);
  }

  /* Returns whereas the field is valid (in a form context) */
  isValid(): boolean {
    return this.props.type === 'single' ? this.state.isSingleDateValid : this.state.isRangeDateValid;
  }

  /* returns the error string, used by date-time-picker to access date errors */
  getErrorString = (): string | undefined => this.state.errorString;

  /* onBlurHandler for NativeInput only (desktop uses airbnb focus methods) */
  onBlurHandler = (): void => {
    this.validateField();
    this.setState({ isFocusReset: true });
  };

  /* Disables/enables next and last month selector buttons based on maximum/minimumDate */
  disableMonthSelectorHandler = (date: moment.Moment): void => {
    const disabledNext = !isNextMonthValid(date, this.props.maximumDate);
    this.setState({ isNextMonthDisabled: disabledNext });
    const disabledPrev = !isPrevMonthValid(date, this.props.minimumDate);
    this.setState({ isPrevMonthDisabled: disabledPrev });
  };

  onChangeVisibleMonthHandler = (date: moment.Moment) => {
    this.setState({ visibleMonth: date });
  };

  render(): JSX.Element {
    const {
      locale,
      className,
      label,
      isLabelHidden,
      requiredLabel,
      optionalLabel,
      subLabel,
      helpButton,
      helpElement,
      isValidationHidden,
      validationErrorRenderer,
      ...partialPropsForDesktop
    } = this.props;

    const { isSingleDateValid, isRangeDateValid, errorString } = this.state;
    const { id, type, isRequired } = partialPropsForDesktop;
    const hasErrors = type === 'single' ? !isSingleDateValid : type === 'range' ? !isRangeDateValid : false;
    const classes = getCSSClasses(
      toolkitstyles.datepicker,
      toolkitstyles['datepicker--haserror'],
      !isValidationHidden,
      hasErrors,
      className
    );
    moment.locale(locale);

    return (
      <div id={`${id}-wrapper`} className={classes}>
        <fieldset>
          {hasErrors && validationErrorRenderer}
          {!isValidationHidden && hasErrors && <ValidationError isValid={!hasErrors} error={errorString ? errorString : 'Error'} />}
          {label && (
            <DateRangePickerLabel
              label={label}
              locale={locale}
              isLabelHidden={isLabelHidden}
              isRequired={isRequired}
              requiredLabel={requiredLabel}
              optionalLabel={optionalLabel}
              helpButton={helpButton}
              subLabel={subLabel}
            />
          )}
          {helpElement ?? null}
          {this.state.isMobile
            ? renderMobileDatePicker(
                this.props,
                this.state,
                this.onSingleDateChange,
                this.onStartDateChange,
                this.onEndDateChange,
                this.onBlurHandler
              )
            : renderDesktopDatePicker(
                partialPropsForDesktop,
                this.state,
                this.airbnbSingleDatepickerRef,
                this.airbnbDateRangepickerRef,
                this.onSingleDateChange,
                this.onSingleDateFocusChange,
                this.onRangeDatesChange,
                this.onRangeDatesFocusChange,
                this.onCalenderIconClick,
                this.onChangeVisibleMonthHandler
              )}
        </fieldset>
      </div>
    );
  }
}

export default DateRangePicker;
