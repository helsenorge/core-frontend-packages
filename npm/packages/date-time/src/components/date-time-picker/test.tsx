import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { render, screen } from '@testing-library/react';
import moment from 'moment';
import { mount, ReactWrapper } from 'enzyme';

import * as DateTimePickerUtils from './date-time-picker-utils';
import ValidationError from '@helsenorge/form/components/form/validation-error';
import DateTimePicker from '.';
import { DateRangePicker } from '../date-range-picker';
import TimeInput from '../time-input';
import { DateTimePickerLegend } from './date-time-picker-legend';
import LanguageLocales from '@helsenorge/core-utils/constants/languages';

describe('DateTimePicker', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  let wrapper: ReactWrapper<{}, {}>;
  const onChangeMock = jest.fn();
  const DateTimeWrapperProps = {
    ref: 'datetime1',
    id: 'datetime1',
    className: 'customclassname',
    legend: 'Når ble du behandlet?',
    dateLabel: 'dato',
    timeLabel: 'klokke',
    initialDate: moment('18.07.2020 12:10', 'DD.MM.YYYY HH:mm'),
    dateValue: moment('18.07.2020', 'DD.MM.YYYY'),
    timeValue: '12:10',
    minimumDateTime: moment('18.07.2020 12:10', 'DD.MM.YYYY HH:mm'),
    maximumDateTime: moment('20.07.2020 22:00', 'DD.MM.YYYY HH:mm'),
    onChange: onChangeMock,
    isRequired: true,
  };

  describe('Når den instansieres for første gang', () => {
    wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
    it('Så renderes den riktig', () => {
      expect(
        wrapper
          .find(DateTimePicker)
          .state('date')
          .format('DD.MM.YYYY')
      ).toEqual('18.07.2020');

      expect(
        wrapper
          .find('fieldset')
          .first()
          .prop('id')
      ).toEqual('datetime1-wrapper');
      expect(
        wrapper
          .find('fieldset')
          .find('div')
          .first()
          .prop('className')
      ).toEqual('mol_validation customclassname');
      expect(wrapper.find(DateTimePickerLegend).length).toEqual(1);
      expect(wrapper.find(DateRangePicker).length).toEqual(1);
      expect(wrapper.find(TimeInput).length).toEqual(1);
      expect(wrapper.find(DateTimePicker).state('time')).toEqual('12:10');
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe('Når DateTimePicker skal vises på samisk', () => {
    it('Så renderes den riktig', () => {
      render(<DateTimePicker id="test" legend="Når ble du behandlet?" locale={LanguageLocales.SAMI_NORTHERN} />);
      const input = screen.getByLabelText('bb.mm.jjjj');
      expect(input).toBeVisible();
    });
  });

  describe('Når value i props oppdateres underveis', () => {
    wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
    it('Så renderes den riktig', () => {
      expect(
        wrapper
          .find(DateTimePicker)
          .state('date')
          .format('DD.MM.YYYY')
      ).toEqual('18.07.2020');
      expect(wrapper.find(DateTimePicker).state('time')).toEqual('12:10');
    });

    it('Så oppdateres den iht incomingProps', () => {
      wrapper.setProps({ dateValue: moment('20.07.2020', 'DD.MM.YYYY'), timeValue: '15:30' });
      expect(
        wrapper
          .find(DateTimePicker)
          .state('date')
          .format('DD.MM.YYYY')
      ).toEqual('20.07.2020');
      expect(wrapper.find(DateTimePicker).state('time')).toEqual('15:30');
    });
  });

  describe('Når den ikke er valid ', () => {
    it('Så renderes den ValidationError i tillegg, med riktig props', () => {
      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
        wrapper.setState({ validated: false, valid: false, errorString: 'errorString' });
        (wrapper.instance() as DateTimePicker).onChildDateValidated();
      });
      expect(
        wrapper
          .find('fieldset')
          .find('div')
          .first()
          .prop('className')
      ).toEqual('mol_validation customclassname mol_validation--active');
      expect(wrapper.find(ValidationError).length).toEqual(1);
      expect(wrapper.find(ValidationError).prop('isValid')).toEqual(false);
      expect(wrapper.find(ValidationError).prop('error')).toEqual('errorString');
    });
  });

  describe('Når isValid kalles', () => {
    it('Så returnerer den riktig isValid state', () => {
      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
        wrapper.setState({ validated: true, valid: true, errorString: undefined });
        (wrapper.instance() as DateTimePicker).onChildDateValidated();
      });

      expect(wrapper.state('valid')).toEqual(true);
      const isValid = (wrapper.instance() as DateTimePicker).isValid();
      expect(isValid).toEqual(true);
      const areAllFieldsValid = (wrapper.instance() as DateTimePicker).areAllFieldsValid();
      expect(areAllFieldsValid).toEqual(true);
    });
  });

  describe('Når validateField kalles', () => {
    const validateMock = jest.fn();
    it('Så kalles det videre validate funksjon', () => {
      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
        (wrapper.instance() as DateTimePicker).validate = validateMock;
        (wrapper.instance() as DateTimePicker).validateField();
      });
      expect(validateMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når validate kalles', () => {
    const isFullDateTimeValidMock = jest.spyOn(DateTimePickerUtils, 'isFullDateTimeValid');

    it('Så sjekkes det om datoen er valid, så kalles det videre notifyValidated funksjon og callback', async () => {
      const cbMock = jest.fn();
      const notifyMock = jest.fn();
      const areAllFieldsValidMock = jest.fn();
      const flushPromises = () => new Promise(setImmediate);

      await act(async () => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);

        wrapper.instance()['timeInputRef'] = {
          current: {
            validateField: () => true,
            isValid: () => true,
          },
        };
        (wrapper.instance() as DateTimePicker).areAllFieldsValid = areAllFieldsValidMock;
        (wrapper.instance() as DateTimePicker).notifyValidated = notifyMock;
        (wrapper.instance() as DateTimePicker).validate(cbMock);
      });
      await flushPromises();
      wrapper.update();

      expect(areAllFieldsValidMock).toHaveBeenCalledTimes(1);
      expect(notifyMock).toHaveBeenCalledTimes(1);
      expect(cbMock).toHaveBeenCalledTimes(1);
      expect(isFullDateTimeValidMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når validateOnlyDate kalles', () => {
    const isFullDateTimeValidMock = jest.spyOn(DateTimePickerUtils, 'isFullDateTimeValid');

    it('Så sjekkes det om datoen er valid, så kalles det videre notifyValidated funksjon og callback', async () => {
      const cbMock = jest.fn();
      const notifyMock = jest.fn();
      const areAllFieldsValidMock = jest.fn();
      const flushPromises = () => new Promise(setImmediate);

      await act(async () => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);

        wrapper.instance()['timeInputRef'] = {
          current: {
            validateField: () => true,
            isValid: () => true,
          },
        };
        (wrapper.instance() as DateTimePicker).areAllFieldsValid = areAllFieldsValidMock;
        (wrapper.instance() as DateTimePicker).notifyValidated = notifyMock;
        (wrapper.instance() as DateTimePicker).validateOnlyDate(cbMock);
      });
      await flushPromises();
      wrapper.update();

      expect(areAllFieldsValidMock).toHaveBeenCalledTimes(0);
      expect(notifyMock).toHaveBeenCalledTimes(1);
      expect(cbMock).toHaveBeenCalledTimes(1);
      expect(isFullDateTimeValidMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når notifyValidated kalles', () => {
    it('Så oppdateres det errorMeldingen og kalles onValidated', () => {
      const updateErrorMessageMock = jest.fn();
      const onValidatedMock = jest.fn();
      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} onValidated={onValidatedMock} />);
        (wrapper.instance() as DateTimePicker).updateErrorMessage = updateErrorMessageMock;
        (wrapper.instance() as DateTimePicker).notifyValidated();
      });
      expect(updateErrorMessageMock).toHaveBeenCalledTimes(1);
      expect(onValidatedMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når onChildDateValidated kalles uten at komponenten er validated', () => {
    it('Så stopper den opp og sjekker den ikke noe validering', () => {
      const isFullDateTimeValidMock = jest.spyOn(DateTimePickerUtils, 'isFullDateTimeValid');

      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
        wrapper.setState({ validated: false });
        (wrapper.instance() as DateTimePicker).onChildDateValidated();
      });
      expect(isFullDateTimeValidMock).toHaveBeenCalledTimes(0);
    });
  });

  describe('Når onChildDateValidated kalles med en tid som er lik den opprinnelige verdien', () => {
    it('Så stopper den opp og  sjekker den ikke noe validering', () => {
      const isFullDateTimeValidMock = jest.spyOn(DateTimePickerUtils, 'isFullDateTimeValid');

      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
        wrapper.setState({ time: '12:10' });
        (wrapper.instance() as DateTimePicker).onChildDateValidated();
      });
      expect(isFullDateTimeValidMock).toHaveBeenCalledTimes(0);
    });
  });

  describe('Når onChildDateValidated kalles uten prop value men med state time', () => {
    const isFullDateTimeValidMock = jest.spyOn(DateTimePickerUtils, 'isFullDateTimeValid');
    const notifyMock = jest.fn();
    const areAllFieldsValidMock = jest.fn();
    it('Så sjekker den validering og kaller notifyValidated videre', async () => {
      const flushPromises = () => new Promise(setImmediate);
      await act(async () => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
        wrapper.instance()['timeInputRef'] = {
          current: {
            isValid: () => true,
          },
        };
        (wrapper.instance() as DateTimePicker).notifyValidated = notifyMock;
        (wrapper.instance() as DateTimePicker).areAllFieldsValid = areAllFieldsValidMock;
        wrapper.setState({ validated: true, time: '18:10' });
      });
      await flushPromises();
      wrapper.update();
      await act(async () => {
        expect(wrapper.find(DateTimePicker).state('time')).toEqual('18:10');
        (wrapper.instance() as DateTimePicker).onChildDateValidated();
      });
      await flushPromises();
      wrapper.update();
      expect(isFullDateTimeValidMock).toHaveBeenCalledTimes(1);
      expect(notifyMock).toHaveBeenCalledTimes(1);
      expect(notifyMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når onDateBlur kalles ', () => {
    it('Så sjekkes det validering (date only)og kalles videre notify og onBlur', () => {
      const onBlurMock = jest.fn();
      const notifyMock = jest.fn();
      const validateOnlyDateMock = jest.fn();
      const getFullMomentDateMock = jest.spyOn(DateTimePickerUtils, 'getFullMomentDate');

      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} onBlur={onBlurMock} />);
        (wrapper.instance() as DateTimePicker).validateOnlyDate = validateOnlyDateMock;
        (wrapper.instance() as DateTimePicker).notifyValidated = notifyMock;
        (wrapper.instance() as DateTimePicker).onDateBlur();
      });
      expect(getFullMomentDateMock).toHaveBeenCalledTimes(1);
      expect(validateOnlyDateMock).toHaveBeenCalledTimes(1);
      expect(validateOnlyDateMock).toHaveBeenCalledWith(notifyMock);
      expect(onBlurMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når onTimeBlur kalles ', () => {
    it('Så sjekkes det validering og kalles videre notify og onBlur', () => {
      const onBlurMock = jest.fn();
      const notifyMock = jest.fn();
      const validateMock = jest.fn();
      const getFullMomentDateMock = jest.spyOn(DateTimePickerUtils, 'getFullMomentDate');

      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} onBlur={onBlurMock} />);
        (wrapper.instance() as DateTimePicker).validate = validateMock;
        (wrapper.instance() as DateTimePicker).notifyValidated = notifyMock;
        (wrapper.instance() as DateTimePicker).onTimeBlur();
      });
      expect(getFullMomentDateMock).toHaveBeenCalledTimes(1);
      expect(validateMock).toHaveBeenCalledTimes(1);
      expect(validateMock).toHaveBeenCalledWith(notifyMock);
      expect(onBlurMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når onTimeChange kalles ', () => {
    it('Så kalles det videre onChange prop', () => {
      const onChangeMock = jest.fn();

      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} onChange={onChangeMock} />);
        (wrapper.instance() as DateTimePicker).onTimeChange('12:10');
      });

      expect(onChangeMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når onDateChange kalles ', () => {
    it('Så kalles det videre onChange prop', () => {
      const onChangeMock = jest.fn();

      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} onChange={onChangeMock} />);
        (wrapper.instance() as DateTimePicker).onDateChange(moment());
      });

      expect(onChangeMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når updateErrorMessage kalles ', () => {
    it('Så kalles det videre onChange prop', () => {
      const getErrorStringMock = jest.spyOn(DateTimePickerUtils, 'getErrorString');

      act(() => {
        wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} />);
        (wrapper.instance() as DateTimePicker).updateErrorMessage();
      });

      expect(getErrorStringMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når den har helpElement', () => {
    it('Så rendres den med helpElement under legend', () => {
      wrapper = mount(<DateTimePicker {...DateTimeWrapperProps} helpElement={<span className="helpElement">{`help`}</span>} />);
      expect(wrapper.find('.helpElement').length).toEqual(1);
    });
  });
});
