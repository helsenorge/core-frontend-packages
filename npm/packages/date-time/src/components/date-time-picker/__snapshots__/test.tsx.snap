// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePicker Når den instansieres for første gang Så renderes den riktig 1`] = `
<fieldset
  id="datetime1-wrapper"
>
  <div
    class="mol_validation customclassname"
  >
    <div
      aria-hidden="false"
      class="ReactCollapse--collapse"
      style="height: auto; overflow: initial;"
    >
      <div
        class="ReactCollapse--content"
      >
        <div
          aria-live="assertive"
          class="mol_validation__errortext"
        />
      </div>
    </div>
    <legend>
      Når ble du behandlet?
    </legend>
    <div
      class="datetimepicker"
    >
      <div
        class="datetimepicker__date"
      >
        <div
          class="datepicker"
          id="datetime1-date-wrapper"
        >
          <fieldset>
            <legend
              class="datepicker__legend"
            >
              <span
                class="datepicker__legend__label "
              >
                dato
              </span>
              <em
                class="datepicker__legend__sublabel"
              >
                 (dd.mm.åååå)
              </em>
            </legend>
            <div
              class="SingleDatePicker SingleDatePicker_1"
            >
              <div>
                <div
                  class="SingleDatePickerInput SingleDatePickerInput_1 SingleDatePickerInput__withBorder SingleDatePickerInput__withBorder_2"
                >
                  <div
                    class="DateInput DateInput_1"
                  >
                    <input
                      aria-describedby="DateInput__screen-reader-message-datetime1-date"
                      aria-label="dd.mm.åååå"
                      autocomplete="off"
                      class="DateInput_input DateInput_input_1"
                      id="datetime1-date"
                      name="datetime1-date"
                      placeholder="dd.mm.åååå"
                      required=""
                      type="text"
                      value="18.07.2020"
                    />
                    <p
                      class="DateInput_screenReaderMessage DateInput_screenReaderMessage_1"
                      id="DateInput__screen-reader-message-datetime1-date"
                    >
                      Skriv inn dato eller tab videre for å bruke kalenderen
                    </p>
                  </div>
                  <button
                    aria-label="Bruk kalenderen og velg dato."
                    class="SingleDatePickerInput_calendarIcon SingleDatePickerInput_calendarIcon_1"
                    type="button"
                  >
                    <span>
                      <svg
                        aria-hidden="true"
                        class="hnds-style-icon datepicker__input-icon"
                        data-analyticsid="icon"
                        fill="black"
                        height="36"
                        role="img"
                        style="min-width: 36px; min-height: 36px;"
                        viewBox="0 0 48 48"
                        width="36"
                      >
                        <path
                          d="M32.86 22.986a.984.984 0 010 1.967.984.984 0 010-1.967zm0 3.03a2.049 2.049 0 002.046-2.047 2.049 2.049 0 00-2.046-2.046 2.049 2.049 0 00-2.046 2.046c0 1.13.918 2.047 2.046 2.047zm6.21-7.864H8.68v-6.194h3.846v1.161a2.243 2.243 0 002.241 2.241 2.243 2.243 0 002.241-2.24v-1.162H30.62v1.161a2.244 2.244 0 002.241 2.241 2.243 2.243 0 002.241-2.24v-1.162h3.97v6.194zM8.678 35.982H39.07v-16.53H8.678v16.53zm5.148-26.273c0-.518.422-.94.941-.94.52 0 .941.422.941.94v3.41a.942.942 0 01-1.882 0V9.71zm18.093 0c0-.518.422-.94.941-.94.52 0 .941.422.941.94v3.41a.942.942 0 01-1.882 0V9.71zm3.182.95v-.95a2.243 2.243 0 00-2.24-2.24 2.244 2.244 0 00-2.242 2.24v.95H17.01v-.95a2.243 2.243 0 00-2.242-2.24 2.243 2.243 0 00-2.24 2.24v.95H7.378V37.28h32.992V10.658h-5.27zm-8.157 16.167a.984.984 0 110 1.969.984.984 0 010-1.969zm0 3.03a2.049 2.049 0 002.047-2.047 2.049 2.049 0 00-2.047-2.046 2.049 2.049 0 00-2.046 2.046c0 1.13.918 2.047 2.046 2.047z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </fieldset>
        </div>
      </div>
      <div
        class="datetimepicker__time"
      >
        <div
          class="mol_timeinput datetimepicker__time__timeinput undefined"
          id="datetime1-time-wrapper"
        >
          <fieldset>
            <legend>
              <span
                class="mol_timeinput__legend"
              >
                <span
                  class="mol_timeinput__legend__label"
                >
                  klokke
                </span>
              </span>
            </legend>
            <div
              class="mol_timeinput__inputs"
            >
              <div
                class="safeInputField"
                id="datetime1-time_hours-wrapper"
              >
                <input
                  aria-label="Timer"
                  aria-required="true"
                  autocomplete="off"
                  class="hn-safe-input atom_input atom_input--medium atom_input--2"
                  max="23"
                  min="0"
                  required=""
                  type="number"
                  value="12"
                />
              </div>
              <span
                class="mol_timeinput__separator"
              >
                :
              </span>
              <div
                class="safeInputField"
                id="datetime1-time_minutes-wrapper"
              >
                <input
                  aria-label="Minutter"
                  aria-required="true"
                  autocomplete="off"
                  class="hn-safe-input atom_input atom_input--medium atom_input--2"
                  max="59"
                  min="0"
                  required=""
                  type="number"
                  value="10"
                />
              </div>
            </div>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;
